rpc: Work around Clang thread safety analysis quirks

practicalswift opened the pull request:
The conditional lock ...

```
LOCK2(cs_main, pwallet ? &pwallet->cs_wallet : nullptr)
```

... confuses Clang's thread safety analysis (it complains about not holding the mutex `pwallet->cs_wallet` even in the case when `pwallet` is non-`NULL`).

So does the access to `pwallet->mapKeyMetadata` via `meta` (it complains about not holding the mutex `cs_wallet` despite `pwallet->cs_wallet` being held).

This commit introduces locking that Clang's thread safety analysis is able to comprehend.

fanquake commented:
This looks like #11691 without the mining.cpp changes. Can you incorporate the changes suggested in comments from that PR?

promag commented on src/rpc/misc.cpp:
```diff
@@ -158,11 +158,8 @@ UniValue validateaddress(const JSONRPCRequest& request)
-#else
-    LOCK(cs_main);
 #endif
+    LOCK(cs_main);
```
How about not changing behaviour:
```cpp
LOCK(cs_main);
#ifdef ENABLE_WALLET
if (pwallet) LOCK(pwallet->cs_wallet);
#endif
```
As it is, the wallet can change between the `cs_main` lock and `cs_wallet` lock. This way the diff is also shorter.

practicalswift replied:
@promag Regarding your edit - do you mean that you ran into Clang thread safety analysis warnings?

promag reviewed (commented):
Concept ACK.

practicalswift commented:
@fanquake Done!

practicalswift commented:
@promag Would you mind re-reviewing? The only way to make the Clang thread safety analysis machinery happy without changing behaviour seems to be to repeat the `LOCK(pwallet->cs_wallet)` (which is OK: `pwallet->cs_wallet` is a recursive mutex). Let me know if you find a cleaner way to do it.

practicalswift committed:
rpc: Work around Clang thread safety analysis quirks

The conditional lock ...

```
LOCK2(cs_main, pwallet ? &pwallet->cs_wallet : nullptr)
```

... confuses Clang's thread safety analysis (it complains
about not holding the mutex pwallet->cs_wallet even in the
case when pwallet is non-NULL).

So does the access to `pwallet->mapKeyMetadata` via `meta`
(it complains about not holding the mutex `cs_wallet`
despite `pwallet->cs_wallet` being held).

This commit introduces locking that Clang's thread safety
analysis is able to understand correctly.

TheBlueMatt commented:
It probably makes more sense to focus more on splitting out the wallet RPC stuff (ala @achow101's work) and then you can just include the wallet lock in a wallet-specific deprecated function.

On November 20, 2017 5:18:15 AM PST, practicalswift <notifications@github.com> wrote:
>The conditional lock ...
>
>```
>LOCK2(cs_main, pwallet ? &pwallet->cs_wallet : nullptr)
>```
>
>... confuses Clang's thread safety analysis (it complains about not
>holding the mutex `pwallet->cs_wallet` even in the case when `pwallet`
>is non-`NULL`).
>
>So does the access to `pwallet->mapKeyMetadata` via `meta` (it
>complains about not holding the mutex `cs_wallet` despite
>`pwallet->cs_wallet` being held).
>
>This commit introduces locking that Clang's thread safety analysis is
>able to comprehend.
>You can view, comment on, or merge this pull request online at:
>
>  https://github.com/bitcoin/bitcoin/pull/11734
>
>-- Commit Summary --
>
>  * rpc: Work around Clang thread safety analysis quirks
>
>-- File Changes --
>
>    M src/rpc/misc.cpp (21)
>
>-- Patch Links --
>
>https://github.com/bitcoin/bitcoin/pull/11734.patch
>https://github.com/bitcoin/bitcoin/pull/11734.diff
>
>-- 
>You are receiving this because you are subscribed to this thread.
>Reply to this email directly or view it on GitHub:
>https://github.com/bitcoin/bitcoin/pull/11734

sipa commented on src/rpc/misc.cpp:
```diff
@@ -158,8 +158,10 @@ UniValue validateaddress(const JSONRPCRequest& request)
-    LOCK2(cs_main, pwallet ? &pwallet->cs_wallet : nullptr);
+    LOCK(cs_main);
+    if (pwallet) {
+        LOCK(pwallet->cs_wallet);
```
This has no effect, the lock is released at the closing brace.

practicalswift replied:
Oh, of course! I don't know what I was thinking there :-\  Fixing!

practicalswift commented:
@sipa Thanks for reviewing! New version pushed. Would you mind re-reviewing? :-)
